### PR TITLE
Ethan - Add effective capacity to List Commons

### LINKED ISSUE #4: FEATURE: As an admin, I can specify a capacity "per farmer"
# User Story

* As an admin
* I can specify a commons capacity on a per farmer basis
* so that the game will scale with the number of users automatically

### Discussion

Prof. de Vries would like to be able to specify commons capacity on a "per user" basis.

The idea is to add a field `capacityPerUser` that will be multiplied by the number of users (farmers) in the commons.    

Then, when doing capacity calculations for the health update formulas, the capacity used will be `max(carryingCapacity, numUsers * capacityPerUser)`.

### Acceptance Criteria

- [x] There is a way for the admin to set capacityPerUser in addition to carryingCapacity; and both fields appear when you create or edit a commons.
- [x] On the `/admin/listcommons` page, you can see both the carrying capacity, the perUserCapacity, and the effectiveCapacity, which is `max(carryingCapacity, numUsers * capacityPerUser)`
- [x] The effectiveCapacity is used in update health formulas instead of the carryingCapacity.

### Implementation Todos

# Fix backend crud operations

- [x] The Commons entity has a new field for capacityPerUser
- [x] The POST operation for creating a new commons can set this field
- [x] The PUT operation for editing a commons can update this field
- [x] The commons entity has a getter for effectiveCapacity which return `max(carryingCapacity, numUsers * capacityPerUser)`
- [x] The commons entity has annotation to return the value of this serialized in the JSON.

# Fix health operations
- [x] The health formulas are updated to use effectiveCapacity

# Fix frontend
- [x] Frontend create and edit pages for commons provide access to capacityPerUser
- [x] List Commons page for admins shows capacityPerUser and effectiveCapacity in addition to all current fields.

### PR DESCRIPTION
**Pull request by EthanSun210**
In this PR, I added effective capacity so that the cows' health is updated according to the effective capacity.

It utilizes the following formula to calculate the effective capacity of the common when doing health calculations
max(capacityPerUser * numUsers, carryingCapacity)

A new column called "Effective  Capacity" will be displayed on the "List Commons" page.
image

Closes: #4

### CODE REVIEW

**Review by haowu0521 — DISMISSED** [review]
LGTM

**Review by GDW1 — CHANGES_REQUESTED** [review]
Update the name to be more specific. I.E. Add to where

I let another PR with a bad name be merged but you should definitely be more specific

**Comment by ZixiaoJin** [comment]
LGTM

**Comment by EthanSun210** [comment]
> 

I've modified the description of the PR.

File: src/main/java/edu/ucsb/cs156/happiercows/controllers/CommonsController.java [hunk 1]
```diff
@@ -221,6 +223,9 @@ public ResponseEntity<String> joinCommon(
 
         userCommonsRepository.save(uc);
 
+        joinedCommons.setNumUsers(joinedCommons.getNumUsers() + 1);
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/controllers/CommonsController.java:226]: This approach is not acceptable.   It adds redundant state to the Commons, and violated database normalization principles.  There is already a method on the commonsRepository for getting the number of  users.  That is how this must be implemented.   Other teams were forced to change their code after implementing this approach.

File: src/main/java/edu/ucsb/cs156/happiercows/entities/Commons.java [hunk 2]
```diff
@@ -44,4 +46,11 @@ public class Commons {
     @OneToMany(mappedBy = "commons", cascade = CascadeType.REMOVE)
     @JsonIgnore
     private List<UserCommons> joinedUsers;
+    @JsonIgnore
+    private int numUsers;
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/entities/Commons.java:50]: This is not permitted.

**Review by pconrad — CHANGES_REQUESTED** [review]
This is not mergeable in its current form, because it uses an incorrect approach, one that sets a trap for future developers.

If you had submitted this earlier, we would have had time to discuss alternative approaches.  
There is very little time left, but if you want to try an alternative, here's an outline.  To be clear, you aren't required to do this, and I'm not really expecting you to, but if you wanted to try, this is how you could go about it.

(1) For the jobs, instead of adding effective capacity to the Commons entity, you can add a function that calculates it from two parameters: a Commons, and a CommonsRepository.  The CommonsRepository has this method to return the number of cows:

https://github.com/ucsb-cs156-m23/proj-happycows-m23-10am-1/blob/3bb2ffeafec58ee9029cefa5c604c5e0ce4ef974/src/main/java/edu/ucsb/cs156/happiercows/repositories/CommonsRepository.java#L17C4-L18C51

So this: 
* `commons.getEffectiveCapacity()`
Would be replaced  by:
* `calculateEffectiveCapacity(commons, commonsRepository)`

You might also need to change the parameters that the health strategies take so that you can get that value into that object.

(2) This may or may not be necessary, but if you need the value to display in the frontend, you can add the effective capacity to the CommonsPlus object here: 

https://github.com/ucsb-cs156-m23/proj-happycows-m23-10am-1/blob/main/src/main/java/edu/ucsb/cs156/happiercows/entities/CommonsPlus.java

This object is initialized in this method of the Commons controller: 

https://github.com/ucsb-cs156-m23/proj-happycows-m23-10am-1/blob/3bb2ffeafec58ee9029cefa5c604c5e0ce4ef974/src/main/java/edu/ucsb/cs156/happiercows/controllers/CommonsController.java#L264

**Comment by pconrad** [comment]
It appears you won't be working on this, so closing it.

**Pull request closed without merging**